File: easy/leetcode_35.cpp
// 35. Search Insert Position
// https://leetcode.com/problems/search-insert-position/

#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <stack>
#include <queue>
#include <utility>

class Solution {
public:
    int searchInsert(std::vector<int>& nums, int target)
    {
        int low = 0, mid = 0, right = nums.size() - 1;

        while (low <= right)
        {
            mid = low + (right - low) / 2;
            if (nums[mid] == target) { return mid; }

            if (nums[mid] < target)
            {
                low = mid + 1;
                if (low > right) return low;
            }
            else if (nums[mid] > target)
            {
                right = mid - 1;
                if (right < low && low == 0) return 0;
            }
        }

        return mid;
    }
};

File: easy/leetcode_35.py
# 35. Search Insert Position
# https://leetcode.com/problems/search-insert-position/

from typing import List

class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        # 7 // 2 == 3
        # 8 // 2 == 4
        left, right = 0, len(nums) - 1
        
        while left <= right:
            # mid = left + (right - left) // 2
            mid = (left + right) // 2

            if nums[mid] == target: 
                return mid
            if nums[mid] < target: 
                left = mid + 1
                if right < left: return left
            elif nums[mid] > target:
                right = mid - 1
                if right < 0: return 0
        
        return mid

s = Solution();
print(s.searchInsert([1, 3], 2))

print(f"1 // 2 = {1 + (1 - 1) // 2}")